examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+	inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultBlitPipelineLayout() const { return blitPipelineLayout; }
+
+	static inline void getDefaultAlphaTestCounterBufferRange(asset::SBufferRange<video::IGPUBuffer>* outBufferRange, core::smart_refctd_ptr<video::IGPUBuffer> scratchBuffer, video::ILogicalDevice* logicalDevice)
+	{
+		if (scratchBuffer)
+		{
+			outBufferRange->offset = 0ull;
+			outBufferRange->size = core::alignUp(sizeof(uint32_t), logicalDevice->getPhysicalDevice()->getLimits().SSBOAlignment);
+			outBufferRange->buffer = scratchBuffer;
+		}
+	}
+
+	static inline void getDefaultAlphaHistogramBufferRange(asset::SBufferRange<video::IGPUBuffer>* outBufferRange, core::smart_refctd_ptr<video::IGPUBuffer> scratchBuffer, video::ILogicalDevice* logicalDevice)
+	{
+		if (scratchBuffer)
+		{
+			outBufferRange->offset = core::alignUp(sizeof(uint32_t), logicalDevice->getPhysicalDevice()->getLimits().SSBOAlignment);
+			outBufferRange->size = sizeof(uint32_t)*NBL_GLSL_DEFAULT_BIN_COUNT;
+			outBufferRange->buffer = scratchBuffer;
+		}
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDefaultAlphaTestDSLayout()
+	{
+		if (!alphaTestDSLayout)
+		{
+			constexpr uint32_t DESCRIPTOR_COUNT = 2u;
+			asset::E_DESCRIPTOR_TYPE types[DESCRIPTOR_COUNT] = { asset::EDT_COMBINED_IMAGE_SAMPLER, asset::EDT_STORAGE_BUFFER }; // input image, alpha test atomic counter
+			alphaTestDSLayout = getDSLayout(DESCRIPTOR_COUNT, types, device, sampler);
+		}
+
+		return alphaTestDSLayout;
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultAlphaTestPipelineLayout()
+	{
+		if (!alphaTestPipelineLayout)
+		{
+			asset::SPushConstantRange pcRange = {};
+			{
+				pcRange.stageFlags = asset::IShader::ESS_COMPUTE;
+				pcRange.offset = 0u;
+				pcRange.size = sizeof(alpha_test_push_constants_t);
+			}
+
+			alphaTestPipelineLayout = device->createGPUPipelineLayout(&pcRange, &pcRange + 1ull, core::smart_refctd_ptr(alphaTestDSLayout));
+		}
+
+		return alphaTestPipelineLayout;
+	}
+
+	core::smart_refctd_ptr<video::IGPUSpecializedShader> createAlphaTestSpecializedShader(const asset::IImage::E_TYPE inImageType)
+	{
+		const char* shaderpath = "../default_compute_alpha_test.comp";
+		auto cpuShader = loadShaderFromFile(shaderpath);
+
+		if (!cpuShader)
+			return nullptr;
+
+		auto cpuShaderOverriden = asset::IGLSLCompiler::createOverridenCopy(cpuShader.get(),
+			"#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d\n"
+			"#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d\n"
+			"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
+			"#define _NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ %d\n",
+			inImageType >= asset::IImage::ET_1D ? NBL_GLSL_DEFAULT_WORKGROUP_DIM : 1u,
+			inImageType >= asset::IImage::ET_2D ? NBL_GLSL_DEFAULT_WORKGROUP_DIM : 1u,
+			inImageType >= asset::IImage::ET_3D ? NBL_GLSL_DEFAULT_WORKGROUP_DIM : 1u,
+			static_cast<uint32_t>(inImageType) + 1u);
+
+		cpuShaderOverriden->setFilePathHint(shaderpath);
```

Comment:
the cleanest way to do this, is to create a micro shader with sprintf or ostream
```glsl
#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d
#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d
#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d
#define _NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ %d

#include "%s"
```

and slap the `shaderpath` as the `%s`, then create GPU shader and specialized shader a s usual.

It should all work, provided `shaderpath` is a built in.

Reply:
still somewhat unresolved.

Btw what is `_NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ ` for?

Reply:
> still somewhat unresolved.

I have yet to do this.

`_NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ ` is for giving `texelFetch` the correct data type (`ivec*`) depending on what `sampler*D` is used.

Reply:
see https://github.com/Devsh-Graphics-Programming/Nabla/pull/343/files/90a68ce14ea09ca939b05feeb27adf38073708c8..4442fb70c98dead6b3be3eff2f6146bd0a466da5#r873930537

Reply:
Resolved.